src/Umbraco.Web.Common/Builder/UmbracoBuilderExtensions.cs

```diff
@@ -5,85 +5,314 @@
+using Umbraco.Infrastructure.HostedServices.ServerRegistration;
+using Umbraco.Web.Common.ApplicationModels;
+using Umbraco.Web.Common.AspNetCore;
 using Umbraco.Web.Common.Filters;
 using Umbraco.Web.Common.ModelBinders;
+using IHostingEnvironment = Umbraco.Core.Hosting.IHostingEnvironment;
 
 namespace Umbraco.Web.Common.Builder
 {
     // TODO: We could add parameters to configure each of these for flexibility
     public static class UmbracoBuilderExtensions
     {
-        public static IUmbracoBuilder AddUmbraco(this IServiceCollection services, IWebHostEnvironment webHostEnvironment, IConfiguration config)
+        public static IUmbracoBuilder AddUmbraco(
+            this IServiceCollection services,
+            IWebHostEnvironment webHostEnvironment,
+            IConfiguration config)
         {
             if (services is null) throw new ArgumentNullException(nameof(services));
-            if (webHostEnvironment is null) throw new ArgumentNullException(nameof(webHostEnvironment));
             if (config is null) throw new ArgumentNullException(nameof(config));
 
-            services.AddLazySupport();
+            var loggingConfig = new LoggingConfiguration(Path.Combine(webHostEnvironment.ContentRootPath, "umbraco", "logs"));
 
-            var builder = new UmbracoBuilder(services, webHostEnvironment, config);
+            var hostingSettings = config.GetSection(Core.Constants.Configuration.ConfigHosting).Get<HostingSettings>() ?? new HostingSettings();
+            var hostingEnvironment = new AspNetCoreHostingEnvironment(new OptionsMonitorAdapter<HostingSettings>(hostingSettings), webHostEnvironment);
+            services.AddLogger(hostingEnvironment, loggingConfig, config);
+
+            IHttpContextAccessor httpContextAccessor = new HttpContextAccessor();
+            services.AddSingleton(httpContextAccessor);
+
+            var requestCache = new GenericDictionaryRequestAppCache(() => httpContextAccessor.HttpContext?.Items);
+            var appCaches = AppCaches.Create(requestCache);
+            services.AddUnique<AppCaches>(appCaches);
+
+            var loggerFactory = LoggerFactory.Create(cfg => cfg.AddSerilog(Log.Logger, false));
+            var typeLoader = services.AddTypeLoader(Assembly.GetEntryAssembly(), webHostEnvironment, hostingEnvironment, loggerFactory, appCaches, config);
+
+            return new UmbracoBuilder(services, config, typeLoader, loggerFactory);
+        }
+
+        /// <remarks>Composes Composers</remarks>
+        public static IUmbracoBuilder AddUmbracoCore(this IUmbracoBuilder builder)
+        {
+            if (builder is null) throw new ArgumentNullException(nameof(builder));
+
+            builder.Services.AddLazySupport();
+
+            // Add service session
+            // This can be overwritten by the user by adding their own call to AddSession
+            // since the last call of AddSession take precedence
+            builder.Services.AddSession(options =>
```

Comment:
It is requeired for macrorendering. So guess it's no problem it it is part of a `AddWebComponents `

Reply:
Moved to AddWebComponents.

It can go anywhere in the Startup.ConfigureServices step as long as it's called before Startup.Configure, this is true for every single registration now.

We don't ever build a service provider, framework does it for us so nothing ever has to be registered in time for something else in the ConfigureServices step.